internal/notebook/resolver.go

```diff
@@ -0,0 +1,132 @@
+package notebook
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/stateful/runme/v3/internal/notebook/daggershell"
+	notebookv1alpha1 "github.com/stateful/runme/v3/pkg/api/gen/proto/go/runme/notebook/v1alpha1"
+	parserv1 "github.com/stateful/runme/v3/pkg/api/gen/proto/go/runme/parser/v1"
+	"github.com/stateful/runme/v3/pkg/document/editor/editorservice"
+	"go.uber.org/zap"
+)
+
+type NotebookResolver struct {
+	notebook *parserv1.Notebook
+	editor   parserv1.ParserServiceServer
+}
+
+func NewNotebookResolver(notebook *parserv1.Notebook) *NotebookResolver {
+	return &NotebookResolver{
+		notebook: notebook,
+		editor:   editorservice.NewParserServiceServer(zap.NewNop()),
+	}
+}
+
+func (r *NotebookResolver) ParseNotebook(context context.Context) (*parserv1.Notebook, error) {
+	// make id sticky only for resolving purposes
+	for _, cell := range r.notebook.Cells {
+		if cell.GetKind() != parserv1.CellKind_CELL_KIND_CODE {
+			continue
+		}
+
+		_, ok := cell.Metadata["id"]
+		if ok {
+			continue
+		}
+
+		if cell.Metadata == nil {
+			return nil, fmt.Errorf("cell metadata is missing")
+		}
+
+		cell.Metadata["id"] = cell.Metadata["runme.dev/id"]
+	}
+
+	// properly parse frontmatter and notebook/cell metadata
+	ser, err := r.editor.Serialize(context, &parserv1.SerializeRequest{Notebook: r.notebook})
+	if err != nil {
+		return nil, err
+	}
+	des, err := r.editor.Deserialize(context, &parserv1.DeserializeRequest{Source: ser.Result})
+	if err != nil {
+		return nil, err
+	}
+
+	return des.Notebook, nil
+}
+
+func (r *NotebookResolver) ResolveDaggerShell(context context.Context, cellIndex uint32) (*notebookv1alpha1.ResolveNotebookResponse, error) {
+	notebook, err := r.ParseNotebook(context)
+	if err != nil {
+		return nil, err
+	}
+
+	var targetCell *parserv1.Cell
+	targetName := ""
+	for idx, cell := range notebook.Cells {
```

Comment:
Suggestion: I think you can skip `for`: 
```suggestion
    // TODO: check if not out of bounds
	cell := notebook.Cells[cellIndex]
```

Reply:
👍